src/main/java/org/hyperledger/fabric/sdk/Endpoint.java

```diff
@@ -61,243 +70,318 @@
+
+        final NettyChannelBuilder channelBuilder = NettyChannelBuilder.forAddress(addr, port)
+                                                       .negotiationType(NegotiationType.PLAINTEXT)
+                                                       .intercept(clientInterceptor);
+        addNettyBuilderProps(channelBuilder, properties);
+        return channelBuilder;
+    }
+
+    private NettyChannelBuilder createGrpcsChannelBuilder(
+        String url, Properties properties, ClientInterceptor clientInterceptor) {
+
+        final NettyChannelBuilder channelBuilder;
+        Optional<Pair<PrivateKey, X509Certificate[]>> clientTLSPropsOpt =
+            readClientTLSProps(url, properties);
+
+        Optional<PrivateKey> clientKey = clientTLSPropsOpt.map(Pair::getKey);
+        Optional<X509Certificate[]> clientCert = clientTLSPropsOpt.map(Pair::getValue);
+
+        Optional<GrpcsContext> grpcsContext = readGrpcsProps(url, properties);
+        Optional<byte[]> pemBytesOpt = grpcsContext.flatMap(GrpcsContext::getPemBytes);
+        Optional<String> cnOpt = grpcsContext.flatMap(GrpcsContext::getCn);
 
-        if (properties != null) {
+        String sslp = readSslpProperty(url, properties);
+        String nt = readNtProperty(url, properties);
 
-            final AbstractMap.SimpleImmutableEntry<PrivateKey, X509Certificate[]> clientTLSProps = getClientTLSProps(properties);
-            if (clientTLSProps != null) {
-                clientCert = clientTLSProps.getValue();
-                clientKey = clientTLSProps.getKey();
+        if (!pemBytesOpt.isPresent()) {
+            // use root certificate
+            channelBuilder = NettyChannelBuilder
+                                 .forAddress(addr, port)
+                                 .intercept(clientInterceptor);
+
+            addNettyBuilderProps(channelBuilder, properties);
+            return channelBuilder;
+        }
+
+        byte[] pemBytes = pemBytesOpt.get();
+        try {
+
+            logger.trace(format("Endpoint %s Negotiation type: '%s', SSLprovider: '%s'",
+                url, nt, sslp
+            ));
+            SslProvider sslprovider = sslp.equals("openSSL") ? SslProvider.OPENSSL : SslProvider.JDK;
+            NegotiationType ntype = nt.equals("TLS") ? NegotiationType.TLS : NegotiationType.PLAINTEXT;
+
+            SslContextBuilder clientContextBuilder = getSslContextBuilder(
+                clientCert.orElse(null), clientKey.orElse(null), sslprovider
+            );
+
+            logger.trace(
+                format("Endpoint %s  final server pemBytes: %s",
+                    url, Hex.encodeHexString(pemBytes)
+                )
+            );
+
+            SslContext sslContext;
+            try (InputStream myInputStream = new ByteArrayInputStream(pemBytes)) {
+                sslContext = clientContextBuilder
+                                 .trustManager(myInputStream)
+                                 .build();
```

Comment:
I think this is the only point in the current code where the IOException (caught at the end of the outer try/catch block) is thrown. If so, the outer try/catch block should be removed and the code here amended as follows:
```suggestion
            try (InputStream myInputStream = new ByteArrayInputStream(pemBytes)) {
                sslContext = clientContextBuilder
                                 .trustManager(myInputStream)
                                 .build();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
```